src/library/scala/concurrent/impl/ExecutionContextImpl.scala

```diff
@@ -53,39 +52,38 @@ private[concurrent] object ExecutionContextImpl {
       thread
     }
 
-    // As per ThreadFactory contract newThread should return `null` if cannot create new thread.
-    def newThread(runnable: Runnable): Thread =
-      if (reserveThread())
-        wire(new Thread(new Runnable {
-          // We have to decrement the current thread count when the thread exits
-          override def run() = try runnable.run() finally deregisterThread()
-        })) else null
+    def newThread(runnable: Runnable): Thread = wire(new Thread(runnable))
 
     def newThread(fjp: ForkJoinPool): ForkJoinWorkerThread =
-      if (reserveThread()) {
-        wire(new ForkJoinWorkerThread(fjp) with BlockContext {
-          // We have to decrement the current thread count when the thread exits
-          final override def onTermination(exception: Throwable): Unit = deregisterThread()
-          final override def blockOn[T](thunk: =>T)(implicit permission: CanAwait): T = {
-            var result: T = null.asInstanceOf[T]
-            ForkJoinPool.managedBlock(new ForkJoinPool.ManagedBlocker {
-              @volatile var isdone = false
-              override def block(): Boolean = {
-                result = try {
-                    // When we block, switch out the BlockContext temporarily so that nested blocking does not created N new Threads
-                    BlockContext.withBlockContext(BlockContext.defaultBlockContext) { thunk }
-                  } finally {
-                    isdone = true
-                  }
-
-                true
-              }
-              override def isReleasable = isdone
-            })
-            result
-          }
-        })
-      } else null
+      wire(new ForkJoinWorkerThread(fjp) with BlockContext {
+        private[this] var isBlocked: Boolean = false // This is only ever read & written if this thread is the current thread
+        final override def blockOn[T](thunk: =>T)(implicit permission: CanAwait): T =
+          if ((Thread.currentThread eq this) && !isBlocked && newBlocker()) {
+            try {
+              isBlocked = true
+              val b: ForkJoinPool.ManagedBlocker with (() => T) =
+                new AtomicInteger(0) with ForkJoinPool.ManagedBlocker with (() => T) {
+                  @volatile private[this] var result: T = null.asInstanceOf[T]
+                  final override def block(): Boolean =
+                    if (compareAndSet(0, 1)) { // Do this to *ensure* that the thunk only gets evaluated once
```

Comment:
Looking at `ForkJoinPool.managedBlock`, it seems like it only ever executes `block()` on the current thread. If so, why does `result` need to be `@volatile`, and a CAS needed to ensure `thunk` is only evaluated once? (If I'm mistaken and `block` is called on multiple threads, then this question is moot.)

Reply:
nvm, I see you've asked Doug this question [here](http://cs.oswego.edu/pipermail/concurrency-interest/2018-April/016387.html)

Reply:
Yeah, I found the documentation a bit unclear, and I didn't want to rely on "the implementation is the specification". Especially since exposing the ManagedBlocker instance to the compensating thread makes a ton of sense to me.

Given his response I'll update the code.

Reply:
Updated.